packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +241,126 @@ def setup_opt_parser
+      query_dns = Resolv::DNS.new(nameserver: [local_ip], search: [zone], ndots: 1)
+      a_record = query_dns.getresource(fqdn, Resolv::DNS::Resource::IN::A)
+      ip = a_record.address.to_s if a_record
+      soa_record = query_dns.getresource(zone, Resolv::DNS::Resource::IN::SOA)
+      serial = soa_record.serial if soa_record
+      {
+        ip: ip,
+        serial: serial,
+      }
+    end
+
+    def parse_zone_file(zone_file)
+      run_cmd('rndc freeze')
+      zone_file_data = File.read(zone_file).split("\n")
+      # find the character A enclosed by any whitespace chars
+      a_record = zone_file_data.find { |line| !(line =~ /\sA\s/).nil? }
+      ip = a_record.split('A').last.strip if a_record
+      serial_line = zone_file_data.find { |line| line.downcase.include?('serial') }
+      serial = serial_line.split('Serial').first.to_i if serial_line
+      {
+        ip: ip,
+        serial: serial
+      }
+    end
+
+    def new_dns_values
+      STDOUT.puts 'gathering DNS data...'
+      local_ip = '127.0.0.1'
+      zone = @scenario_answers['foreman_proxy']['dns_zone']
+      reverse_zone = @scenario_answers['foreman_proxy']['dns_reverse']
+      if reverse_zone.is_a?(Array)
+        reverse_zone = reverse_zone.first
+      end
+      soa_admin_domain = "root.#{zone}"
+      key_file = @scenario_answers['foreman_proxy']['keyfile']
+      begin
+        STDOUT.puts 'querying local DNS server...'
+        ip_serial_data = query_dns_server(@old_hostname, zone, local_ip)
+        ip = ip_serial_data[:ip]
+        serial = ip_serial_data[:serial]
+        new_serial = serial + 1 if serial
+        reverse_serial = query_dns_server(@old_hostname, reverse_zone, local_ip)[:serial]
+        new_reverse_serial = reverse_serial + 1 if reverse_serial
+      rescue Resolv::ResolvError => e
+        STDOUT.puts e
+        STDOUT.puts 'Parsing zone file as fallback'
+        ip_serial_data = parse_zone_file("/var/named/dynamic/db.#{zone}")
+        ip = ip_serial_data[:ip]
+        serial = ip_serial_data[:serial]
+        new_serial = serial + 1 if serial
+        reverse_serial = parse_zone_file("/var/named/dynamic/db.#{reverse_zone}")[:serial]
+        new_reverse_serial = reverse_serial + 1 if reverse_serial
+      ensure
+        run_cmd('rndc thaw')
+      end
+
+
+      {
+        local_ip: local_ip,
+        zone: zone,
+        old_fqdn: @old_hostname,
+        soa_admin_domain: soa_admin_domain,
+        ip: ip,
+        new_serial: new_serial,
+        new_reverse_serial: new_reverse_serial,
+        new_fqdn: @new_hostname,
+        key_file: key_file,
+        reverse_zone: reverse_zone
+      }
+    end
+
+    def nsupdate_command(dns_entries, key_file)
+      "echo -e \"#{dns_entries}\" | nsupdate -l -k #{key_file}"
+    end
+
+    def update_dns_records(new_dns_values = {})
+      new_dns_values.each do |key, value|
+        unless value
+          raise "Error gathering DNS data: couldn't find value for '#{key}'"
```

Comment:
its nice to have an actionable when things go wrong. What do you expect the user to do here or where should they check?

Reply:
@johnpmitsch Thanks, updated